filters/FarthestPointSamplingFilter.cpp

```diff
@@ -0,0 +1,128 @@
+ *       names of its contributors may be used to endorse or promote
+ *       products derived from this software without specific prior
+ *       written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+ * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+ * OF SUCH DAMAGE.
+ ****************************************************************************/
+
+#include "FarthestPointSamplingFilter.hpp"
+
+#include <pdal/KDIndex.hpp>
+#include <pdal/util/ProgramArgs.hpp>
+
+#include <algorithm>
+#include <limits>
+#include <numeric>
+#include <string>
+#include <vector>
+
+namespace pdal
+{
+
+static PluginInfo const s_info{"filters.fps", "Farthest point sampling filter",
+                               "http://pdal.io/stages/filters.fps.html"};
+
+CREATE_STATIC_STAGE(FarthestPointSamplingFilter, s_info)
+
+std::string FarthestPointSamplingFilter::getName() const
+{
+    return s_info.name;
+}
+
+FarthestPointSamplingFilter::FarthestPointSamplingFilter()
+{
+}
+
+void FarthestPointSamplingFilter::addArgs(ProgramArgs& args)
+{
+    args.add("count", "Target number of points after sampling", m_count,
+             point_count_t(1000));
+}
+
+PointViewSet FarthestPointSamplingFilter::run(PointViewPtr inView)
+{
+    // Return empty PointViewSet if the input PointView has no points.
+    PointViewSet viewSet;
+    if (!inView->size() || (inView->size() < m_count))
+        return viewSet;
+
+    // Otherwise, make a new output PointView.
+    PointViewPtr outView = inView->makeNew();
+
+    // Construct a KD-tree of the input view.
+    KD3Index kdi(*inView);
+    kdi.build();
+
+    // Seed the output view with the first point in the current sorting.
+    PointId seedId(0);
+    outView->appendPoint(*inView, seedId);
+
+    // Compute distances from seedId to all other points.
+    PointIdList indices(inView->size());
+    std::vector<double> sqr_dists(inView->size());
+    kdi.knnSearch(seedId, inView->size(), &indices, &sqr_dists);
+
+    // Sort distances by PointId.
+    std::vector<double> min_dists(inView->size());
+    for (PointId i = 0; i < inView->size(); ++i)
+        min_dists[indices[i]] = sqr_dists[i];
+
+    // Proceed until we have m_count points in the output PointView.
+    for (PointId i = 1; i < m_count; ++i)
```

Comment:
I guess I don't understand what's being done with min_dists.

Reply:
So the key thing happening with `min_dists` is

https://github.com/PDAL/PDAL/blob/c678bff41306818d99e1d7adc8975c7ad065600b/filters/FarthestPointSamplingFilter.cpp#L119-L120

Keep in mind that from the initialization of `min_dists` above, we had a distance of 0.0 for the `seedId`. Within this loop, we will also have a distance of 0.0 at `idx` (i.e., the most recently added furthest point). By keeping the minimum distance, we now have two (in the first iteration) 0.0 distance entries, one for each point in our output view, meaning they cannot be repeatedly detected as the furthest point, and are removed from further consideration. It also means that of the remaining neighbors, we are now storing the current closest distance to any point that has been added to the output view, which is critical when picking the next furthest point.

Reply:
Same with this.  Sorry.